File: src/main/java/com/example/chatService/security/UserPrincipal.java
package com.example.chatService.security;

import com.example.chatService.entity.AuthProvider;
import com.example.chatService.entity.User;
import lombok.Getter;
import org.springframework.security.core.GrantedAuthority;
import org.springframework.security.core.authority.SimpleGrantedAuthority;
import org.springframework.security.core.userdetails.UserDetails;
import org.springframework.security.oauth2.core.user.OAuth2User;

import java.util.Collection;
import java.util.Collections;
import java.util.List;
import java.util.Map;

@Getter
public class UserPrincipal implements UserDetails, OAuth2User {

    private final String id;
    private final String email;
    private final String username;
    private final String password;
    private final boolean nicknameCompleted;
    private final AuthProvider provider;
    private final String providerId;
    private final Collection<? extends GrantedAuthority> authorities;
    private Map<String, Object> attributes;

    private UserPrincipal(
            String id,
            String email,
            String username,
            String password,
            boolean nicknameCompleted,
            AuthProvider provider,
            String providerId,
            Collection<? extends GrantedAuthority> authorities
    ) {
        this.id = id;
        this.email = email;
        this.username = username;
        this.password = password;
        this.nicknameCompleted = nicknameCompleted;
        this.provider = provider;
        this.providerId = providerId;
        this.authorities = authorities;
    }

    public static UserPrincipal from(User user) {
        List<GrantedAuthority> authorities = Collections.singletonList(
                new SimpleGrantedAuthority("ROLE_" + user.getRole())
        );
        return new UserPrincipal(
                user.getId(),
                user.getEmail(),
                user.getUsername(),
                user.getPassword(),
                user.isNicknameCompleted(),
                user.getProvider(),
                user.getProviderId(),
                authorities
        );
    }

    // JWT 일반로그인
    public static UserPrincipal create(User user) {
        List<GrantedAuthority> authorities =
                List.of(new SimpleGrantedAuthority("ROLE_USER"));

        return new UserPrincipal(
                user.getId(),
                user.getEmail(),
                user.getUsername(),
                user.getPassword(),
                user.isNicknameCompleted(),
                user.getProvider(),
                user.getProviderId(),
                authorities
        );
    }

    // OAuth2.0 로그인
    public static UserPrincipal create(User user, Map<String, Object> attributes) {
        UserPrincipal principal = create(user);
        principal.attributes = attributes;
        return principal;
    }

    @Override
    public Map<String, Object> getAttributes() {
        return attributes;
    }

    @Override
    public String getName() {
        return email != null && !email.isBlank()
            ? email
            : provider.name().toLowerCase() + "_" + providerId;
    }

//    @Override
//    public String getName() {
//        return id;
//    }

    public String getId() {
        return String.valueOf(id);
    }

    /* ================= UserDetails ================= */

    @Override
    public boolean isAccountNonExpired() {
        return true;
    }

    @Override
    public boolean isAccountNonLocked() {
        return true;
    }

    @Override
    public boolean isCredentialsNonExpired() {
        return true;
    }

    @Override
    public boolean isEnabled() {
        return true;
    }
}

File: src/main/java/com/example/chatService/security/UserPrincipal.kt
package com.example.chatService.security

import com.example.chatService.entity.AuthProvider
import com.example.chatService.entity.User
import org.springframework.security.core.GrantedAuthority
import org.springframework.security.core.authority.SimpleGrantedAuthority
import org.springframework.security.core.userdetails.UserDetails
import org.springframework.security.oauth2.core.user.OAuth2User

class UserPrincipal private constructor(
    val id: String,
    private val email: String?,
    private val username: String?,
    private val password: String?,
    val isNicknameCompleted: Boolean,
    val provider: AuthProvider?,
    val providerId: String?,
    private val authorities: Collection<GrantedAuthority>,
    private var attributes: Map<String, Any>? = null
) : UserDetails, OAuth2User {

    override fun getAuthorities(): Collection<GrantedAuthority> = authorities
    override fun getPassword(): String? = password
    override fun getUsername(): String? = username
    override fun isAccountNonExpired(): Boolean = true
    override fun isAccountNonLocked(): Boolean = true
    override fun isCredentialsNonExpired(): Boolean = true
    override fun isEnabled(): Boolean = true
    override fun getAttributes(): Map<String, Any>? = attributes

    override fun getName(): String {
        return if (!email.isNullOrBlank()) email
        else "${provider?.name?.lowercase()}_$providerId"
    }

    companion object {
        @JvmStatic
        fun from(user: User): UserPrincipal {
            val authorities = listOf(SimpleGrantedAuthority("ROLE_${user.role}"))
            return UserPrincipal(
                id = user.id ?: "",
                email = user.email,
                username = user.username,
                password = user.password,
                isNicknameCompleted = user.nicknameCompleted,
                provider = user.provider,
                providerId = user.providerId,
                authorities = authorities
            )
        }

        @JvmStatic
        fun create(user: User): UserPrincipal {
            val authorities = listOf(SimpleGrantedAuthority("ROLE_USER"))
            return UserPrincipal(
                id = user.id ?: "",
                email = user.email,
                username = user.username,
                password = user.password,
                isNicknameCompleted = user.nicknameCompleted,
                provider = user.provider,
                providerId = user.providerId,
                authorities = authorities
            )
        }

        @JvmStatic
        fun create(user: User, attributes: Map<String, Any>): UserPrincipal {
            return create(user).also { it.attributes = attributes }
        }
    }
}
